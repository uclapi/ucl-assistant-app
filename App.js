import { AppLoading, Notifications } from "expo"
import { Asset } from "expo-asset"
import * as Font from "expo-font"
import PropTypes from "prop-types"
import React, { Component } from "react"
import { Platform, StatusBar, View } from "react-native"
import { Provider } from "react-redux"
import { PersistGate } from "redux-persist/es/integration/react"

import configureStore from "./configureStore"
import Colors from './constants/Colors'
import { NotificationChannels } from "./constants/notificationsConstants"
import { AnalyticsManager, AssetManager, ErrorManager } from "./lib"
import RootNavigation from "./navigation/RootNavigation"
import Styles from "./styles/Containers"

const { persistor, store } = configureStore

ErrorManager.initialise()

if (process.env.NODE_ENV !== `production`) {
  const whyDidYouRender = require(`@welldone-software/why-did-you-render`)
  whyDidYouRender(React)
}

class App extends Component {
  static propTypes = {
    skipLoadingScreen: PropTypes.bool,
  }

  static defaultProps = {
    skipLoadingScreen: false,
  }

  static getActiveRouteName(navigationState) {
    if (!navigationState) {
      return null
    }
    const route = navigationState.routes[navigationState.index]
    // dive into nested navigators
    if (route.routes) {
      return App.getActiveRouteName(route)
    }
    return route.routeName
  }

  constructor(props) {
    super(props)
    this.state = {
      isLoadingComplete: false,
      store: {
        persistor,
        store,
      },
    }
  }

  componentDidMount() {
    if (Platform.OS === `android`) {
      Object.keys(NotificationChannels).forEach((key) => {
        const channel = NotificationChannels[key]
        Notifications.createChannelAndroidAsync(channel.id, channel.options)
      })
    }
    this.notificationSubscription = Notifications.addListener(
      this.handleNotification,
    )
    AnalyticsManager.initialise()
  }

  loadResourcesAsync = async () => Promise.all([
    Asset.loadAsync(Object.values(AssetManager.undraw)),
    Font.loadAsync({
      ...AssetManager.font,
    }),
  ])

  handleLoadingError = (error) => {
    ErrorManager.captureError(error)
  }

  handleFinishLoading = () => {
    this.setState({ isLoadingComplete: true })
  }

  handleNotification = (notification) => console.log(
    `Received notification`,
    notification,
  )

  onNavigationStateChange = (prevState, currentState) => {
    const currentScreen = App.getActiveRouteName(currentState)
    const prevScreen = App.getActiveRouteName(prevState)

    if (prevScreen !== currentScreen) {
      AnalyticsManager.logScreenView(currentScreen)
    }
  }

  render() {
    const { isLoadingComplete } = this.state
    const { skipLoadingScreen } = this.props
    if (!isLoadingComplete && !skipLoadingScreen) {
      return (
        <AppLoading
          startAsync={this.loadResourcesAsync}
          onError={this.handleLoadingError}
          onFinish={this.handleFinishLoading}
        />
      )
    }
    const {
      store: { store: stateStore, persistor: statePersistor },
    } = this.state
    return (
      <Provider store={stateStore}>
        <PersistGate persistor={statePersistor}>
          <View style={Styles.app}>
            <StatusBar
              barStyle="dark-content"
              hidden={false}
              backgroundColor={Colors.pageBackground}
            />
            <RootNavigation
              onNavigationStateChange={this.onNavigationStateChange}
            />
          </View>
        </PersistGate>
      </Provider>
    )
  }
}

export default App
