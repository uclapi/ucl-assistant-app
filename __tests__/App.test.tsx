import { render } from "@testing-library/react-native"
import React from "react"
import "react-native"
import App from "../App"

jest.mock(`react-redux`, () => ({
  Provider: ({ children }) => children,
  connect: () => (component) => component,
}))

jest.mock(`redux-persist/lib/integration/react`, () => ({
  PersistGate: ({ children }) => children,
}))

jest.mock(`../configureStore`, () => ({
  persistor: jest.fn(),
  store: jest.fn(),
}))

describe(`App`, () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.useRealTimers()
  })

  it(`renders the loading screen`, async () => {
    const tree = render(<App />)
    expect(tree).toMatchSnapshot()
  })
})
