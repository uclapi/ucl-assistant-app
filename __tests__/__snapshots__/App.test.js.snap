// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the loading screen 1`] = `null`;

exports[`renders the root without loading screen 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fefefe",
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      style={
        Object {
          "backgroundColor": "#eee",
          "flex": 1,
          "flexDirection": "column-reverse",
          "overflow": "hidden",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          pointerEvents="auto"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
                Object {
                  "translateY": 0,
                },
              ],
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            style={
              Object {
                "backgroundColor": "#fff",
                "flex": 1,
              }
            }
          >
            <ViewManagerAdapter_ExpoLinearGradient
              proxiedProperties={
                Object {
                  "colors": Array [
                    4279998859,
                    4283151357,
                  ],
                  "endPoint": Array [
                    1,
                    0,
                  ],
                  "locations": undefined,
                  "startPoint": Array [
                    0,
                    1,
                  ],
                }
              }
              style={
                Array [
                  Object {
                    "backgroundColor": "#fefefe",
                    "flex": 1,
                    "paddingLeft": 20,
                    "paddingRight": 20,
                  },
                  Object {
                    "paddingBottom": 20,
                    "paddingTop": 60,
                  },
                ]
              }
            >
              <RCTSafeAreaView
                emulateUnlessSupported={true}
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <Image
                  resizeMethod="scale"
                  resizeMode="contain"
                  source={1}
                  style={
                    Object {
                      "alignSelf": "stretch",
                      "backgroundColor": "rgba(0,0,0,0)",
                      "flex": 1,
                      "height": undefined,
                      "marginTop": 50,
                      "width": undefined,
                    }
                  }
                />
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#4D5061",
                        "fontFamily": "apercu-bold",
                        "fontSize": 28,
                        "marginBottom": 2,
                        "marginTop": 8,
                      },
                      Object {
                        "color": "#fefefe",
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  One app to manage your life at UCL.
                </Text>
                <View
                  style={
                    Object {
                      "flex": 1,
                    }
                  }
                />
                <View
                  accessible={true}
                  isTVSelectable={true}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "opacity": 1,
                    }
                  }
                >
                  <ViewManagerAdapter_ExpoLinearGradient
                    proxiedProperties={
                      Object {
                        "colors": Array [
                          4279998859,
                          4283151357,
                        ],
                        "endPoint": Array [
                          1,
                          0,
                        ],
                        "locations": undefined,
                        "startPoint": Array [
                          0,
                          1,
                        ],
                      }
                    }
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "borderRadius": 8,
                          "elevation": 2,
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "marginBottom": 5,
                          "marginTop": 5,
                          "minHeight": 55,
                          "minWidth": 150,
                          "padding": 8,
                          "paddingLeft": 16,
                          "paddingRight": 16,
                          "shadowColor": "#1B998B",
                          "shadowOffset": Object {
                            "height": 2,
                            "width": 0,
                          },
                          "shadowOpacity": 0.75,
                          "shadowRadius": 3,
                        },
                        Object {
                          "borderColor": "#fefefe",
                          "borderWidth": 2,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "alignItems": "center",
                            "flexDirection": "row",
                            "justifyContent": "center",
                          },
                          Object {},
                        ]
                      }
                    >
                      <Image
                        resizeMethod="scale"
                        resizeMode="contain"
                        source={1}
                        style={
                          Array [
                            Object {
                              "alignSelf": "stretch",
                              "backgroundColor": "rgba(0,0,0,0)",
                              "flex": 1,
                              "height": undefined,
                              "marginTop": 50,
                              "width": undefined,
                            },
                            Object {
                              "alignSelf": "center",
                              "flexBasis": 20,
                              "flexGrow": 0,
                              "height": 20,
                              "marginRight": 5,
                              "marginTop": 0,
                            },
                          ]
                        }
                      />
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#fefefe",
                              "fontFamily": "apercu",
                              "fontSize": 20,
                              "marginBottom": 1,
                              "marginTop": 1,
                              "textAlign": "center",
                            },
                            Object {},
                          ]
                        }
                      >
                        Sign In With UCL
                      </Text>
                    </View>
                  </ViewManagerAdapter_ExpoLinearGradient>
                </View>
                <View
                  style={
                    Object {
                      "marginTop": 20,
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#4D5061",
                          "fontFamily": "apercu",
                          "fontSize": 16,
                          "marginBottom": 1,
                          "marginTop": 1,
                        },
                        Object {},
                      ]
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#4D5061",
                            "fontFamily": "apercu",
                            "fontSize": 16,
                            "marginBottom": 1,
                            "marginTop": 1,
                          },
                          Object {
                            "color": "#fff",
                          },
                        ]
                      }
                    >
                      By signing into this app, you agree toÂ 
                    </Text>
                    <Text
                      onPress={[Function]}
                      style={
                        Array [
                          Object {
                            "color": "#4D5061",
                            "fontFamily": "apercu",
                            "fontSize": 16,
                            "marginBottom": 1,
                            "marginTop": 1,
                          },
                          Object {
                            "color": "#2980b9",
                          },
                          Object {
                            "color": "#fff",
                            "textDecorationLine": "underline",
                          },
                        ]
                      }
                    >
                      UCL API's terms & conditions.
                    </Text>
                  </Text>
                </View>
              </RCTSafeAreaView>
            </ViewManagerAdapter_ExpoLinearGradient>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "transform": Array [
                Object {
                  "translateX": -750,
                },
              ],
            }
          }
        >
          <View
            onLayout={[Function]}
            pointerEvents="box-none"
            style={
              Object {
                "backgroundColor": "#fff",
                "borderBottomColor": "#a7a7aa",
                "borderBottomWidth": 0.5,
                "height": 64,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 20,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
